### PR TITLE
Issue 5360 - Fix styles on block tone change

### LINKED ISSUE #5360: Bug with colours due to changing tone
### Description

Some of blocks, SVG, texts are not changing colours after changing tone ( only after click to it )

### Step-by-step reproduction instructions

1) load SML-PRO-162 ( for example )
2) change tone
3) check colours for all svg and texts

### Screenshots, screen recording, code snippet

https://jam.dev/c/a12deb02-128c-4e30-a6cb-2bd050b12df2

### URL and/or Code editor

_No response_

### How to fix it

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by Fybex**
# Description

Fixes #5360 

# How Has This Been Tested?

# Test checklist

**_ Front/Back Testing _**

-   [x] Test if all blocks in container change styles on block tone change
-   [x] Test if block changes tone(and styles) on moving from one tone container to another
-   [x] Test if it's not slower in comparison with master

**_ Final-Code Testing _**

-   [ ] Test if all blocks in container change styles on block tone change
-   [ ] Test if block changes tone(and styles) on moving from one tone container to another
-   [ ] Test if it's not slower in comparison with master, if it's optimized
-   [ ] Check no commented code and no unnecessary imports
-   [ ] Standards of the project have been followed
-   [ ] No errors/warnings on console

# Checklist

-   [X] My code follows the style guidelines of this project
-   [X] I have performed a self-review of my code
-   [X] My changes generate no new warnings/errors

## Summary by CodeRabbit

- **Enhancements**
  - Streamlined the `block-styles-control` by optimizing the `getAllInnerBlocks` function and removing redundant code.
  - Improved state management in the `withAttributes` extension using `useSelect` and `useDispatch` hooks for more efficient attribute handling.

- **Refactor**
  - Optimized logic for determining block hierarchy and style application in the `withAttributes` extension.

- **Bug Fixes**
  - Addressed issues related to conditional attribute updates for a more reliable editing experience.

- **Code Clean-up**
  - Simplified the codebase by removing the unused `isEmpty` import.

### CODE REVIEW

File: src/extensions/attributes/withAttributes.js [hunk 1]
```diff
@@ -118,6 +114,30 @@ const withAttributes = createHigherOrderComponent(
...
+					if (blockStyle !== attributes.blockStyle) {
+						markNextChangeAsNotPersistent();
+						setAttributes({
+							blockStyle,
+						});
+					}
+				}
+
+				if (isFirstOnHierarchy !== attributes.isFirstOnHierarchy) {
+					attributes.isFirstOnHierarchy = isFirstOnHierarchy;
```
[the remark this answers is not in the document]
Comment (Fybex) [reply src/extensions/attributes/withAttributes.js]: it's expected in our case, we don't want to cause block re-render on `isFirstOnHierarchy` change, only on `blockStyle` change to update styles

**Review by Olekrut — CHANGES_REQUESTED** [review]
1) looks like blocks which should be fixed, are rendering slower
https://jam.dev/c/a221fa51-472a-45fe-b07c-4ff2f7e8ec5c
2) when we are moving blocks from one tone to another, colours are different
https://jam.dev/c/82c1d4be-77f5-425e-8b50-ce6c471d87b6

**Comment by Fybex** [comment]
> 1. looks like blocks which should be fixed, are rendering slower
>    https://jam.dev/c/a221fa51-472a-45fe-b07c-4ff2f7e8ec5c
> 2. when we are moving blocks from one tone to another, colours are different
>    https://jam.dev/c/82c1d4be-77f5-425e-8b50-ce6c471d87b6

1. There's a lot of blocks in that pattern, it's natural to be a bit slower than before, because now it rerenders this blocks 👌 There's definitely some area for improvement, but it's not that fast to improve, so going to leave it for later
2. All good here actually 👌 . Other arrows have black on the white tone, this is their problem, we move the column where the icon had the first colour from the palette, it remains the first, but the palette changes to light, and the colour accordingly

**Comment by Fybex** [comment]
The average execution time on block tone change was significantly reduced from approximately 3245.62 milliseconds to 1016.91 milliseconds (tested on SML-PRO-162), marking an 68.67% improvement.

Test it please @Olekrut 🙏. It uses style caching approach, which was introduced in #4698 for breakpoint changes

**Comment by Olekrut** [comment]
YY, looks like tone changes a bit faster now

**Review by Olekrut — APPROVED** [review]

**Review by elzadj — CHANGES_REQUESTED** [review]
Backend works fine, but it's still Light on the frontend. @Olekrut could you check it too please?

[screencast-wordpress.local-2024.01.03-13_36_05.webm](https://github.com/maxi-blocks/maxi-blocks/assets/2401618/d6b5f3e2-0343-4e6c-90b1-b10974d2fb98)

**Comment by Olekrut** [comment]
YY, i have this bug as well, probably it happened after optimization

> Backend works fine, but it's still Light on the frontend. @Olekrut could you check it too please?
> 
>  screencast-wordpress.local-2024.01.03-13_36_05.webm

**Review by Olekrut — APPROVED** [review]

**Review by elzadj — APPROVED** [review]

**Pull request merged**